Add "Latest Posts" Block

lamosty opened the pull request:
@mtias suggested to me to create a simple block to get latest posts with the WP REST API. It should also serve as an example for other developers so they can start experimenting with the API and create some cool blocks.

I've been studying Gutenberg (pretty good editor!) and come up with something really simple. It's still a work in progress since I want to fix a few things and maybe make it a bit better.

I'd like to also open a discussion about some of the things here:
- naming: I prefixed it with `rest-api-` since we could have another 'latest posts' block without using rest api
- new category called `rest-api`: are we planning to have more such components using rest api? If yes, I think it would be nice to give them their own category
- getting the posts/using the rest api: I'm using the [Backbone rest api js client](https://developer.wordpress.org/rest-api/using-the-rest-api/backbone-javascript-client/) as it seems to be the officially supported one. It offers a few niceties and it's quite easy to use but: isn't Backbone a bit dated? (I've never used it). Are we okay with this or should we create some custom rest api wrapper/client?

Asking for help: after saving a post with this latest posts block and then reloading the saved post, instead of having the block parsed and displayed as other blocks, it shows with the Gutenberg tags and HTML code. Can you spot what I'm doing wrong, please? It works fine when seeing the post on the front-end though.

Am I using the `attributes` of `registerBlock` properly? I use them to store the latest posts retrieved from DB but don't define them in `registerBlock`.

Thanks!

## Nice to have/TODO
- [ ] have some UI so users can filter the posts by number of them to show, author, categories, etc

lamosty committed:
Add initial scaffolding for REST API Latest Posts block

lamosty committed:
Load 5 latest posts into unordered list

youknowriad commented:
Thanks for doing this. 

At the moment, this block will shows the latest post at the moment where we save the block, if we create a new post after that. This new post is not included in the frontend. I think this post is a good opportunity to leverage the Server Side Rendering to address this.

lamosty commented:
> I think this post is a good opportunity to leverage the Server Side Rendering to address this.

Yeah, I've seen your PR but haven't looked into it more. I also agree since it's an expected behavior to always display the latest posts, especially when using REST API (ie it's 'dynamic').

mtias commented:
Thanks for giving this a try!

> I also agree since it's an expected behavior to always display the latest posts, especially when using REST API (ie it's 'dynamic').

Definitely, this is a great example of a purely dynamic block in my book.

mtias commented on blocks/library/rest-api-latest-posts/index.js:
```diff
@@ -0,0 +1,71 @@
+ */
+import { registerBlock } from '../../api';
+
+function getLatestPosts( postsToShow = 5 ) {
```
Thinking... this could be a good example for breaking some logic into a separate file. Perhaps `rest-api-latest-posts/data.js`?

mtias commented on blocks/library/rest-api-latest-posts/index.js:
```diff
@@ -0,0 +1,71 @@
+/**
```
Let's call this block just `latest-posts`.

westonruter commented:
There's good overlap here with the Recent Posts widget.

lamosty committed:
Refactor to `latest-posts` instead of `rest-api-latest-posts`

jasmussen commented:
Sweet deal, and awesome work!

I was so inspired I took a stab at the design of this, as suggested in your ToDo. Note: the following is very mockuppy and conceptual, and I expect some pushback here. The implementation is also probably something that can happen best in a separate PR, but nonetheless to keep the discussion grouped for now, here it is. 

Neutral:

![recent posts neutral](https://cloud.githubusercontent.com/assets/1204802/26344758/010f4ca2-3fa1-11e7-90a8-f944be9c8267.png)

Selected:

![recent posts selected](https://cloud.githubusercontent.com/assets/1204802/26344954/a707ee3e-3fa1-11e7-8117-92c44edde5b8.png)

This leverages both the "block-first" UI of putting the most basic stuff right on the block, and the more advanced stuff in the Inspector on the right. For this concept, I arbitrarily decided to put the post date and number of posts in the inspector, but this was also done in order to grease the discussion on what we put there (see #672 ).

lamosty committed:
Move data fetchers to data.js file

Props @mtias for the suggestion

mtias commented:
@jasmussen I like this, but I think we could also avoid it looking like a form if we just made the title an optional editable field like caption/quote-source with placeholder text instead of a label and an input.

jasmussen commented:
> I like this, but I think we could also avoid it looking like a form if we just made the title an optional editable field like caption/quote-source with placeholder text instead of a label and an input.

This was my first instinct as well. I do think we'll want some further iteration here, as to what feels right. But I specifically like the gray background which is a callback to placeholders. Not only does this help the block feel distinct from the others, in that it's dynamic, but it also literally "grays" the content, implying you can't edit it directly.

lamosty commented:
Question: can `attributes` contain an object? For example, I'm storing the retrieved posts as JS objects in `attributes` and upon calling `save`, they get rendered into the comment as:

![selection_302](https://cloud.githubusercontent.com/assets/4988512/26348606/80c69f70-3fad-11e7-85fc-06518467151b.png)

However, it seems like Gutenberg doesn't like if attribute is `"[object Object]"` and doesn't render  the block properly when loading a saved instance.

youknowriad commented:
@lamosty `attributes` should not store objects (At least it can't for now). I'd suggest you store the posts in a local state (the `edit` function can be a React Component). And I'd just return `null` in the save function right now.

lamosty committed:
Refactor edit() to render a React component

Fixes the issue of having to use `attributes` to store the latest posts and is also nicer.

lamosty committed:
Add poststoshow attribute with a default value 5

lamosty committed:
Add server-side rendering

I also had to rename the block from core/latest-posts to
core/latestposts since I think there's a bug in the server-side matcher
-- it ignores block comments with hyphens: #882

lamosty commented:
In 497cee8, I added simple server-side rendering. However, I had to rename the component to `core/latestposts` from `core/latest-posts` because if there's a hyphen in the name, the server-side parser will ignore it. I think that's a bug and opened a new issue: #882.

Now, I have some more questions :):
- when doing server-side rendering of a block, we have to write the render logic again and try to match it as closely as possible to how it works in JS. Are there some plans/ideas to add ability to write just one render logic and have it render on both client and server?
- is it possible to use the same code that runs in the editor to render stuff on site's front-end? I.e. use the `edit` method to work on front-end? If not, are there any plans to support this?

youknowriad commented:
> when doing server-side rendering of a block, we have to write the render logic again and try to match it as closely as possible to how it works in JS. Are there some plans/ideas to add ability to write just one render logic and have it render on both client and server?

Yes, @notnownikki proposed a SSR endpoint see (https://github.com/WordPress/gutenberg/issues/780) and implemented it in the WIP pr #816 
I think this endpoint should be extracted to its own PR

> is it possible to use the same code that runs in the editor to render stuff on site's front-end? I.e. use the edit method to work on front-end? If not, are there any plans to support this?

For now, the frontend is just a raw HTML, saved from the editor (the `save`function can share some of the `edit` elements) or Server side rendered. I guess you're proposing to use a React component as an output, It's the first time I'm seeing this proposal, it raises so many questions, maybe worth creating an issue to discuss this possibility (but I don't see this for a v1 personally)

youknowriad commented on blocks/library/latest-posts/index.js:
```diff
@@ -0,0 +1,82 @@
+
+				this.props.setAttributes( {
+					poststoshow
+				} );
```
I haved mixed feelings here, we're calling `setAttributes` in the constructor which could change the markup directly. I think it's better to use `defaultAttributes` property for this (see usage in the `table` block I think).

lamosty replied:
Ah, I didn't know we have `defaultAttributes`. Thanks!

notnownikki replied:
I'm happy to put the SSR endpoint into its own PR, can start on it now if it's wanted :)

lamosty replied:
> I'm happy to put the SSR endpoint into its own PR, can start on it now if it's wanted :)

No rush, I think I'll leave the SSR part as it is now and we can improve it in a future PR. :)

paaljoachim commented:
Here are the various options that are used in the Genesis featured posts widget. I am posting it to give additional information on options that might be good to include. 

![genesis-featured-posts](https://cloud.githubusercontent.com/assets/5323259/26418333/21424aee-40bc-11e7-83ed-2d12d022d20e.jpg)

lamosty commented:
> Here are the various options that are used in the Genesis featured posts widget. I am posting it to give additional information on options that might be good to include.

Wow, that's quite a list!

I think I'll try to keep this simple though and try to implement @jasmussen's design for now. We can always improve. :)

paaljoachim commented:
Yeah too many options in there. That list is good to have as some of the options might be good to add at a later date.

westonruter commented on blocks/library/latest-posts/data.js:
```diff
@@ -0,0 +1,18 @@
+ * Returns a Promise with the latest posts or an error on failure.
+ *
+ * @param   {Number} postsToShow
+ * @returns {Object}
```
Instead of `{Object}` this can be `{wp.api.collections.Posts}`

westonruter commented on blocks/library/latest-posts/index.php:
```diff
@@ -0,0 +1,37 @@
+	foreach( $recent_posts as $post ) {
+		$post_permalink = get_permalink( $post['ID'] );
+
+		$posts_content .= "<li><a href='{$post_permalink}'>{$post['post_title']}</a></li>\n";
```
Instead of `$post['post_title']` this should use `get_the_title( $post['ID'] )` so that filters apply.

westonruter commented on index.php:
```diff
@@ -473,3 +476,8 @@ function the_gutenberg_project() {
 	<?php
 }
+
+function gutenberg_load_blocks_server_side_redering() {
```
Typo `s/redering/rendering/`

westonruter commented on blocks/library/latest-posts/index.php:
```diff
@@ -0,0 +1,37 @@
+<?php
+
+function gutenberg_block_core_latest_posts( $attributes ) {
+	$postsToShow = 5;
```
Shouldn't this be `$posts_to_show`?

lamosty committed:
Use defaultProps to initialize postsToShow

lamosty committed:
Add various fixes according to the review

lamosty committed:
Fix linting errors

nylen commented:
After #849 which added a bunch of tests for block parsing and serialization, this PR should probably be rebased against `master`, either because the build will fail upon merge, or because this PR can add some new tests using this structure ([details here](https://github.com/WordPress/gutenberg/blob/484c146da24c8dac8232452f559161d58023c5fd/blocks/test/fixtures/README.md)).

lamosty committed:
Rebase and fix eslint/WP coding standards

youknowriad commented on blocks/library/latest-posts/index.js:
```diff
@@ -0,0 +1,75 @@
+import { getLatestPosts } from './data.js';
+
+registerBlockType( 'core/latestposts', {
+	title: wp.i18n.__( 'Latest Posts' ),
```
Minor: Should we be consistent about importing `wp.*` and use an `import { __ } from 'i18n';` here?

lamosty replied:
I think yes. However, how can I import `wp.*`? `wp` is a global variable which is already present so I'm not sure how I would import it.

youknowriad replied:
`import { __ } from 'i18n';` This would work, all the `wp.*` modules defined in Gutenberg are configured to allow importing them like this.

lamosty replied:
Ah, perfect, thanks!

youknowriad commented on blocks/library/latest-posts/index.js:
```diff
@@ -0,0 +1,75 @@
+/**
+ * WordPress dependencies
+ */
+import Placeholder from 'components/placeholder';
```
I think we should use `import { Placeholder } from 'components';` to avoid code duplication when bundling the modules.

youknowriad commented on blocks/library/latest-posts/index.js:
```diff
@@ -0,0 +1,75 @@
+
+			const { poststoshow } = this.props.attributes;
+
+			getLatestPosts( poststoshow )
```
We should probably abort the query if the component is unmounted before getting the response (to avoid React warnings)

lamosty replied:
Fixed in 64b30f1.

However, I'm not sure how I can easily test/trigger `componentWillUnmount`. Do you have some idea? :)

youknowriad replied:
Is there any reason to store the request in the state. I tend to use simple class properties when changes to this property don't have to trigger a rerender.

> However, I'm not sure how I can easily test/trigger componentWillUnmount. Do you have some idea? :)

yep, hard to test. I'd suggest throttling the network requests and removing the block before the response (and ensure no warnings are shown in the console)

youknowriad commented on blocks/library/latest-posts/index.js:
```diff
@@ -0,0 +1,75 @@
+				.then( latestPosts => this.setState( { latestPosts } ) );
+		}
+
+		renderPostsLoading() {
```
Personal opinion, feel free to ignore. 

I prefer to avoid these render* functions when they are simple enough. It makes the component easier to refactor IMO (transform to functional or anything else when needed).

aduth replied:
And in the case they're not simple, it's a good sign the component could do to be broken up into multiple smaller self-contained components.

lamosty committed:
Add fn which loads server-side rendering of blocks

This was lost in last rebase

lamosty committed:
Avoid small render* functions and put the whole rendering in render

lamosty committed:
Add full post content fixture

lamosty commented:
In c9c7146, I added a basically blank full post content fixtures. since latest posts block returns `null` in `save`, it doesn't save anything in database. This is what I added in `core-latestposts.html` file:

```
<!-- wp:core/latestposts poststoshow="5" --><!-- /wp:core/latestposts -->
```

But that doesn't test anything (except the `poststoshow` attribute). Should we instead use the HTML which gets rendered from the server-side rendering?

lamosty committed:
Abort the latest posts query if component is unmounted before response

lamosty committed:
Import __ from i18n instead of using wp.i18n.__

youknowriad reviewed (approved):
Aside avoiding to store the request in the state, this looks solid to me 👍

jasmussen commented:
Nice work! 🎉

westonruter commented on blocks/library/latest-posts/index.php:
```diff
@@ -0,0 +1,51 @@
+	$posts_to_show = 5;
+
+	if ( array_key_exists( 'poststoshow', $attributes ) ) {
+		$posts_to_show = $attributes['poststoshow'];
```
Since there is no sanitization or validation of block attributes (yet), it should probably be done here, ensuring it is a number and that it is greater than 0, but less than (maybe) 100. 

See also https://github.com/WordPress/gutenberg/issues/886#issuecomment-305518238

lamosty committed:
Use class prop instead of React state to store posts request

lamosty committed:
Add basic posts to show attribute validation

lamosty commented:
Thank you very much for all the great tips and reviews! @youknowriad @notnownikki @aduth @mtias @westonruter @jasmussen 

I've learned quite a lot here and refreshed my JS skills. :)

mtias commented:
Thank you @lamosty for working on this one.

nylen commented on gutenberg.php:
```diff
@@ -9,6 +9,8 @@
  * @package gutenberg
  */
 
+define( 'GUTENBERG__PLUGIN_DIR', plugin_dir_path( __FILE__ ) );
```
These `define`s are duplicating code already in place elsewhere:  https://github.com/WordPress/gutenberg/blob/3f727da24ee9059759d4ab4669a6681839219a49/lib/client-assets.php#L13-L22

The `GUTENBERG__` (double underscore) convention is unique to the constants introduced here; others in the codebase do not use this convention.  It seems unnecessary to me.

I don't have a preference on whether this is done as a function or constants (the function is pretty fast), but it needs to be consistent in approach and naming.

Also, `gutenberg.php` should contain the bare minimum necessary to initialize the plugin, because it is going to be rewritten during the build process.  These defines should probably happen in a new `lib/constants.php` or similar.

lamosty replied:
Ah, I didn't notice that, my bad.

In a new PR, I'll fix those things.

Thanks!

nylen commented on lib/blocks.php:
```diff
@@ -128,3 +130,12 @@ function do_blocks( $content ) {
+function gutenberg_load_blocks_server_side_rendering() {
+	require_once GUTENBERG__BLOCKS_LIBRARY_DIR . '/latest-posts/index.php';
+}
+add_action( 'init', 'gutenberg_load_blocks_server_side_rendering' );
```
I think there are a few other issues here, not least that `npm run package-plugin` is currently broken because this file was not added to the plugin zip package.  I'll open a PR to try to fix.
